core/performance_test/peakflops/01_cpp_simple.cpp

```diff
@@ -0,0 +1,206 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "common.h"
+/*#include "rdtsc.h"*/
+
+#include <gtest/gtest.h>
+/*
+For skylake
+Model name:            Intel(R) Xeon(R) Gold 6152 CPU @ 2.10GHz
+gcc version 7.2.0
+g++ -O3  -march=native peakflops.cpp -DVECLENTH=16
+640000000.000000 flops
+6482486 clocks
+98.727556 flops/clock
+
+For knl
+Model name:            Intel(R) Xeon Phi(TM) CPU 7250 @ 1.40GHz
+g++ version 8.1.0
+g++ -O3  -march=native peakflops.cpp -DVECLENTH=16
+640000000.000000 flops
+10900834 clocks
+58.711104 flops/clock
+
+For haswell/broadwell
+Model name:            Intel(R) Xeon(R) CPU E5-2660 v3 @ 2.60GHz
+gcc 7.3.0/intel 17.0.6/clang 5.0.1
+
+$ g++ -O3  -march=native  peakflops.cpp -fopt-info-vec
+$ ./a.out
+160000000000.000000 flops
+4200302774 clocks
+38.092492 flops/clock
+x_[0] = 2.161056
+x_[1] = 11.734885
+x_[2] = 4.641315
+x_[3] = 160.230576
+x_[4] = 1.043288
+x_[5] = 6.865434
+x_[6] = 14.501436
+x_[7] = 1.216459
+
+$ icpc -O3 -march=core-avx2 -fma peakflops.cpp
+[gchen@cn151 openmp]$ ./a.out
+160000000000.000000 flops
+4200379130 clocks
+38.091800 flops/clock
+x_[0] = 2.161056
+x_[1] = 11.734885
+x_[2] = 4.641316
+x_[3] = 160.230591
+x_[4] = 1.043288
+x_[5] = 6.865434
+x_[6] = 14.501434
+x_[7] = 1.216459
+
+$ clang++ -O3 peakflops.cpp -mavx2 -mfma -ffp-contract=fast
+$ ./a.out
+160000000000.000000 flops
+4200978198 clocks
+38.086368 flops/clock
+x_[0] = 2.161056
+x_[1] = 11.734885
+x_[2] = 4.641315
+x_[3] = 160.230576
+x_[4] = 1.043288
+x_[5] = 6.865434
+x_[6] = 14.501436
+x_[7] = 1.216459
+-------------------
+For Sandy/Ivy Bridge you need to unroll by 3:
+
+Only FP Add has dependency on the previous iteration of the loop
+FP Add can issue every cycle
+FP Add takes three cycles to complete
+Thus unrolling by 3/1 = 3 completely hides the latency
+FP Mul and FP Load do not have a dependency on the previous iteration and you can rely on the OoO core to issue them in the near-optimal order. These instructions could affect the unroll factor only if they lowered the throughput of FP Add (not the case here, FP Load + FP Add + FP Mul can issue every cycle).
```

Comment:
So you do three "loops" within one?

Reply:
Do you do this unrolling in this file?

Reply:
Honestly, not sure. This is a note from @guangyechen . But I agree I don't see an unroll by 3? I assume it's hidden in the `n`

Reply:
Those are some old note taken from a few years ago. The number of "unrolling" is very cpu dependent, and the note may need be cleaned up.

Reply:
@sslattery To answer your question: yes, here we do unrolling of 10 for haswell cpus.

Reply:
Updated comments in file